src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -5,51 +5,202 @@
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = httpServerProvider.createHttpServer(addrIPv4, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV4)) {
+                throw e;
+            }
+            logger.atInfo().cause(e).kv("address", addrIPv4)
+                    .log("Unable to bind HTTP server. "
+                            + "Proceeding anyway as IPv4 is likely unavailable on the host system");
+        }
+
+        InetSocketAddress addrIPv6 =
+                new InetSocketAddress(ADDR_IPV6,
+                        // use the same port from IPv4 server
+                        serverIPv4 == null ? port : serverIPv4.getAddress().getPort());
+        try {
+            serverIPv6 = httpServerProvider.createHttpServer(addrIPv6, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV6)) {
+                stop();
+                throw e;
+            }
+            if (serverIPv4 == null) {
+                // unable to create either server. likely indicative of a larger issue,
+                // such as no ports being available.
+                throw e;
+            }
+            logger.atInfo().cause(e).kv("address", addrIPv6)
+                    .log("Unable to bind HTTP server. "
+                            + "Proceeding anyway as IPv6 is likely unavailable on the host system");
+        }
+    }
+
+    private boolean canBindToAddress(String hostname) {
+        try (Socket s = socketFactory.createSocket()) {
+            s.bind(new InetSocketAddress(hostname, 0));
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
     }
 
     @Override
     public void start() {
-        httpImpl.createContext(URL, credentialRequestHandler);
-        httpImpl.setExecutor(executorService);
-        httpImpl.start();
+        getServers().forEach(this::start);
+    }
+
+    private void start(HttpServer server) {
+        server.createContext(URL, credentialRequestHandler);
+        server.setExecutor(executorService);
+        server.start();
     }
 
     @Override
     public void stop() {
-        httpImpl.stop(TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS);
+        getServers().forEach(this::stop);
+    }
+
+    private void stop(HttpServer server) {
+        server.stop(TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS);
+    }
+
+    List<InetSocketAddress> getServerAddresses() {
+        return getServers().stream()
+                .map(HttpServer::getAddress)
+                .collect(Collectors.toList());
     }
 
     int getServerPort() {
-        return httpImpl.getAddress().getPort();
+        return getServers().stream()
+                .findFirst()
+                .map(HttpServer::getAddress)
+                .map(InetSocketAddress::getPort)
+                .orElseThrow(() -> new IllegalStateException("No servers created"));
```

Comment:
illegal state is a bad exception. Use checked exceptions or return a reasonable default value.